File: python/creational/abstract_factory.py
class Animal:
    def speak(self):
        pass


class Sound:
    def make_sound(self):
        pass


class Dog(Animal):
    def speak(self):
        print("i'm dog")


class Bark(Sound):
    def make_sound(self):
        print('Woof!')


class Cat(Animal):
    def speak(self):
        print("i'm cat")


class Meow(Sound):
    def make_sound(self):
        print('meow')


class AnimalFactory:
    def createAnimal(self):
        pass

    def createSound(self):
        pass


class DogFactory(AnimalFactory):
    def createAnimal(self):
        return Dog()

    def createSound(self):
        return Bark()


class CatFactory(AnimalFactory):
    def createAnimal(self):
        return Cat()

    def createSound(self):
        return Meow()


def client(factory: AnimalFactory):
    animal = factory.createAnimal()
    sound = factory.createSound()

    animal.speak()
    sound.make_sound()

client(CatFactory())
client(DogFactory())

File: js/creational/abstract_factory.js
class Animal {
    speak() {
    }
}

class AnimalSound {
    makeSound() {
    }
}

class Cat extends Animal {
    speak() {
        console.log("i'm cat")
    }
}

class Meow extends AnimalSound {
    makeSound() {
        console.log('meow')
    }
}

class Dog extends Animal {
    speak() {
        console.log("i'm dog")
    }
}

class Bark extends AnimalSound {
    makeSound() {
        console.log("Woof!")
    }
}

class AnimalFactory {
    createAnimal() {
    }

    createSound() {
    }
}

class CatFactory extends AnimalFactory {
    createAnimal() {
        return new Cat()
    }

    createSound() {
        return new Meow()
    }
}


class DogFactory extends AnimalFactory {

    createAnimal() {
        return new Dog()
    }

    createSound() {
        return new Bark()
    }
}

const client = (factory) => {
    const animal = factory.createAnimal()
    const sound = factory.createSound()

    animal.speak()
    sound.makeSound()
}
client(new DogFactory())
client(new CatFactory())